src/ast.rs

```diff
@@ -0,0 +1,92 @@
+pub struct Item {
+    pub is_pub: Boolean,
+    pub ident: String,
+    pub kind: ItemKind,
+}
+
+pub enum ItemKind {
+    Fn(Fn),
+    Struct,
+    Enum,
+    Mod,
+    Use,
+    Trait,
+    TyAlias,
+}
+
+pub struct Fn {
+    pub generics: Generics,
+    pub sig: FnSig,
+}
+
+pub struct Generics {
+    pub params: Vec<GenericParam>,
+    pub where_clause: Vec<GenericParam>,
+}
+
+pub struct GenericParam {
+    pub ident: String,
+    pub bounds: Vec<GenericBound>,
+    pub kind: GenericParamKind,
+}
+
+pub enum GenericBound {
+    Trait(Path),
+}
+
+pub struct Path {
+    pub segments: Vec<PathSegment>,
+}
+
+pub struct PathSegment {
+    pub ident: String,
+    pub args: Option<GenericArgs>,
+}
+
+pub struct GenericArgs {
+    pub args: Vec<GenericArg>,
+}
+
+pub enum GenericArg {
+    Type(Ty),
+    // Const(AnonConst),
+}
+
+pub struct Ty {
+    pub kind: TyKind,
+}
+
+pub enum TyKind {
+    Path(Path),
+}
+
+pub enum GenericParamKind {
+    Type,
+    // Const,
+}
+
+pub struct FnSig {
+    pub header: FnHeader,
+    pub decl: FnDecl,
+}
+
+pub struct FnHeader {
```

Comment:
Please include safety, constness and ext, with comments as well.
```rust
/// A function header.
///
/// All the information between the visibility and the name of the function is
/// included in this struct (e.g., `async unsafe fn` or `const extern "C" fn`).
#[derive(Clone, Copy, Encodable, Decodable, Debug)]
pub struct FnHeader {
    /// Whether this is `unsafe`, or has a default safety.
    pub safety: Safety,
    /// Whether this is `async`, `gen`, or nothing.
    pub coroutine_kind: Option<CoroutineKind>,
    /// The `const` keyword, if any
    pub constness: Const,
    /// The `extern` keyword and corresponding ABI string, if any.
    pub ext: Extern,
}
```

Reply:
As mentioned above, also accepted your opinion.